Lib/test/test_abc.py

```diff
@@ -467,8 +467,8 @@ def __subclasshook__(cls, C):
                     if cls is A:
                         return 'foo' in C.__dict__
                     return NotImplemented
-            self.assertFalse(issubclass(A, A))
-            self.assertFalse(issubclass(A, (A,)))
+            self.assertTrue(issubclass(A, A))
```

Comment:
This indicates that previously it was possible for classes to customise whether they would be considered subclasses of themselves, but that with this patch, it would no longer be possible for them to do that. Unfortunately I think that counts as a backwards-incompatible behaviour change, so I'm not sure we can do this

Reply:
I think we can do this, because of:
1. The class is a subclass to itself, this is logical. Also, this is written in the documentation
2. This behaviour is similar to 'isinstance'
3. There is very little code that can potentially break, since metaprogramming (especially use of '__subclasscheck__' override) is rarely used

Reply:
> 1. The class is a subclass to itself, this is logical.

The fact that there are several tests asserting this behaviour strongly suggests to me that the current behaviour is intentional. And the current behaviour makes more logical sense to me than the new behaviour you're proposing here. Under the current behaviour, if `__subclasscheck__` is defined for a class, it is _always_ called when `issubclass()` is called against that class; your patch creates a special case where it is not called, which seems confusing and inconsistent.

Even if this _wasn't_ deliberate behaviour -- we're not starting from scratch here, unfortunately. I think it's highly likely that some code out there is relying on being able to customise _all_ `issubclass()` checks against a class, _even_ `issubclass()` checks of a class against itself.

> 1. this is written in the documentation: [docs.python.org/3/library/functions.html#issubclass](https://docs.python.org/3/library/functions.html#issubclass)

Sure, but [it also says in the data model regarding `__subclasscheck__`](https://docs.python.org/3/reference/datamodel.html#class.__subclasscheck__):

> class.__subclasscheck__(self, subclass)
>
>    Return true if `subclass` should be considered a (direct or indirect) subclass of `class`. If defined, called to implement `issubclass(subclass, class)`.

It doesn't say anything about the method being ignored if `subclass` is the exact same class as `class`.

> 3\. There is very little code that can potentially break, since metaprogramming (especially use of `__subclasscheck__` override) is rarely used

I think you might be surprised at how often metaprogramming is used in libraries. Anyway, if I understand correctly, your patch only speeds things up for classes with custom metaclasses, correct? I think `__subclasscheck__` is probably quite commonly overridden for classes with custom metaclasses.

Reply:
Yes, my patch only speeds things up for classes with custom metaclasses

Can you make a hint what I suppose to fix to make this patch backwards compatible?

Reply:
> Can you make a hint what I suppose to fix to make this patch backwards compatible?

We can't take the fast path if the class defines a custom `__subclasscheck__` method, because then it might be customising `issubclass()` to return `False` even when it's comparing itself to itself

Reply:
Maybe, some kind of c api `hasattr` can help?

Reply:
Maybe https://docs.python.org/3/c-api/object.html#c.PyObject_HasAttr?

Reply:
> Maybe, some kind of c api `hasattr` can help?

No, becase `type.__subclasscheck__` exists -- so the metaclass will almost certainly _have_ a `__subclasscheck__` method. You need to check whether the `__subclasscheck__` method on the metaclass is _the same as_ `type.__subclasscheck__`